emacs.d/config.org
#+TITLE: Emacs Init File!
#+PROPERTY: header-args :results silent :comments link :tangle ~/dotfiles/emacs.d/init.el

* Intro

This is the bulk of my emacs configuration, loaded by [[./init.el][init.el]]. It's pretty cool
that I can write it in org mode, because easy organization and rearrangement
are sweet. Check out [[https://github.com/aaronbieber/dotfiles/tree/master/configs/emacs.d][@aaronbieber]], [[https://github.com/hrs/dotfiles/tree/master/emacs.d][@hrs]], and [[https://github.com/syl20bnr/spacemacs][spacemacs]] - a lot of the
inspiration (and erm, code) for this is taken from them.

:NOTE:
*Note:* The usual way to do this is to call ~(org-babel-load-file "config.org")~,
and I did that for a while, but for whatever reason I decided I wanted
=config.el= (and its compiled verison) to be produced as I changed the file
itself, hence the [[#dotfile-handling][dotfile handling]] functions and [[#local-vars][file local variables]]. Also,
the optional ~COMPILE~ argument of ~org-babel-load-file~ wasn't working for me, so
there's that.
:END:

First off, let's set lexical binding, in accordance with the
[[https://github.com/bbatsov/emacs-lisp-style-guide#source-code-layout--organization][emacs-lisp-style-guide]].
#+begin_src emacs-lisp :comments no
;;; -*- lexical-binding: t -*-
;;(package-initialize)
(defconst d/emacs-start-time (current-time))
#+end_src

* Contents                                                            :toc_1:
- [[#intro][Intro]]
- [[#initial-setup][Initial Setup]]
- [[#package-management][Package management]]
- [[#core][Core]]
- [[#help][Help]]
- [[#files-and-buffers][Files and Buffers]]
- [[#editing][Editing]]
- [[#navigation][Navigation]]
- [[#git-and-github][Git and github]]
- [[#appearance][Appearance]]
- [[#modes][Modes]]
- [[#enhancements][Enhancements]]
- [[#fun][Fun]]
- [[#web][Web?]]
- [[#finalize][Finalize]]

* Initial Setup
** Personal Info
#+begin_src emacs-lisp
(setq user-full-name "Diego A. Mundo"
      user-mail-address "diegoamundo@gmail.com")
#+end_src
** Server
Start server if not already running. You can do this with ~emacs --daemon~ or
even automate it with ~brew services start emacs~ on macOS, but I usually just
run Emacs on login anyway, so this suffices.

This makes startup time irrelevant. Start emacs once, connect with emacsclient
every other time. See [[https://www.gnu.org/software/emacs/manual/html_node/emacs/Emacs-Server.html#Emacs-Server][Using Emacs as a Server]].
#+begin_src emacs-lisp
(require 'server)
(unless (server-running-p)
  (server-start))
#+end_src
** Load Path
Add [[./lisp][lisp]] directory and subdirectories to ~load-path~ and ~custom-theme-load-path~.

This is where I put lisp that isn't necessarily central to my config or needs
to be in a standalone file such as a some auth settings, lisp practice,
really bad self-made themes, ~org-export-async-init-file~, etc.
#+begin_src emacs-lisp
(eval-and-compile
  (add-to-list 'load-path (expand-file-name "lisp" user-emacs-directory))
  (let ((default-directory  "~/.emacs.d/lisp/"))
    (normal-top-level-add-subdirs-to-load-path))

  (add-to-list 'custom-theme-load-path "~/.emacs.d/lisp/"))
#+end_src
** cl-lib
I'm sure this will be required by some package somewhere along the way, but I
use this a fair bit so let's explicitly require it.
#+begin_src emacs-lisp
(require 'cl-lib)
#+end_src
* Package management
** [[https://github.com/raxod502/straight.el#getting-started][straight]]
#+begin_src emacs-lisp
(setq straight-recipe-overrides '((nil (straight :host github
                                                 :repo "raxod502/straight.el"
                                                 :files ("straight.el")
                                                 :branch "develop"))))
(let ((bootstrap-file (concat user-emacs-directory "straight/bootstrap.el"))
      (bootstrap-version 2))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
        (url-retrieve-synchronously
         "https://raw.githubusercontent.com/raxod502/straight.el/develop/install.el"
         'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))
#+end_src
** [[https://github.com/jwiegley/use-package][use-package]]
#+begin_quote
  A use-package declaration for simplifying your .emacs
#+end_quote
An excellent utility for managing packages and package configuration in a neat
and organized way, with advanced support for deferring, pre/post-loading
configuration, time reporting, and more.

Using use-package, you can use the same init file across computers without
keeping track of what's installed or not and it will ensure that any missing
packages are installed. It's pretty neat.
#+begin_src emacs-lisp
(straight-use-package 'use-package)
#+end_src
I also set some variables - in particular, tell me the load time when a package
takes more than .001 seconds, always tell me about package loading, and always
defer and ensure packages are installed unless otherwise stated.
#+begin_src emacs-lisp
(setq use-package-minimum-reported-time .001
      use-package-verbose t
      use-package-always-defer t
      use-package-always-ensure t)

(eval-when-compile
  (require 'use-package))
#+end_src
* Core
** Defaults
*** Prompts
Having to type "yes" can be annoying.
#+begin_src emacs-lisp
(defalias 'yes-or-no-p #'y-or-n-p)
#+end_src
*** Messages
Allow more messages in ~*Messages*~ buffer so you can look at what happened waaay
back if you need to.
#+begin_src emacs-lisp
(setq message-log-max 10000)
#+end_src
*** Disabled Commands
#+begin_src emacs-lisp
(setq disabled-command-function nil)
#+end_src
*** Scrolling
Scroll one line at a time, and only scroll the current line when moving past
right boundary.
#+begin_src emacs-lisp
(setq scroll-step 1
      scroll-conservatively 10000
      auto-hscroll-mode 'current-line)
#+end_src
Smoother mouse scrolling, which is now irrelevant to me since I've disabled the
mouse in emacs.
#+begin_src emacs-lisp :tangle no
(setq mouse-wheel-scroll-amount '(2 ((shift) . 1) ((control) . nil))
      mouse-wheel-progressive-speed nil)
#+end_src
*** Minibuffer
Allow editing in the minibuffer... /with/ the minibuffer. Also resize minibuffer
windows to fit text.
#+begin_src emacs-lisp
(setq enable-recursive-minibuffers t
      resize-mini-windows t)
#+end_src
*** Column
Show the column number in the modeline, because I'm not a savage.
#+begin_src emacs-lisp
(setq column-number-mode t)
#+end_src
*** Tab
Use tab for completion and cycling candidates (relevant for the minibuffer?).
#+begin_src emacs-lisp
(setq tab-stop-list (number-sequence 4 200 4)
      completion-cycle-threshold t
      tab-always-indent 'complete)
#+end_src
*** Tramp
Use ssh by default and remember passwords for [[https://www.gnu.org/software/tramp/][tramp]]. Also make it quieter
except for warnings and errors.
#+begin_src emacs-lisp
(setq tramp-default-method "ssh"
      tramp-verbose 2
      password-cache t
      password-cache-expiry 86400)
#+end_src
*** Kill-ring
Save stuff you've copied in other applications to the emacs kill-ring.
#+begin_src emacs-lisp
(setq save-interprogram-paste-before-kill t)
#+end_src
*** Custom
#+begin_src emacs-lisp
(setq custom-file "~/.emacs.d/custom.el")
(load custom-file)
#+end_src
** Functions
*** Dotfiles
#+begin_src emacs-lisp
(defvar d/show-async-tangle-results nil)

(defun d/async-tangle-init ()
  "Asynchronously tangle config.org and and byte-compile."
  (interactive)
  (let ((init-tangle-start-time (current-time)))
    (async-start
     (lambda ()
       ;; make async emacs aware of packages (for byte-compilation)
       (require 'org)

       ;; tangle
       (org-babel-tangle-file
        (expand-file-name "~/dotfiles/emacs.d/config.org")))
     (unless d/show-async-tangle-results
       `(lambda (result)
          (if result
              (message "SUCCESS: config.org successfully tangled. (%.3fs)"
                       (float-time (time-subtract (current-time)
                                                  ',init-tangle-start-time)))
            (message "ERROR: config.org tangle failed.")))))))

(defun d/async-babel-tangle ()
  "Tangle org file asynchronously."
  (interactive)
  (let ((file (buffer-file-name)))
    (async-start
     `(lambda ()
        (require 'org)
        (org-babel-tangle-file ,file))
     (unless d/show-async-tangle-results
       (lambda (result)
         (if result
             (message "SUCCESS: successfully tangled file.")
           (message "ERROR: tangle file failed.")))))))
#+end_src
*** Config helpers
#+begin_src emacs-lisp
(defmacro d/setup-mode (modes &rest body)
  "Create a setup function for MODES and add it to relevant hook(s)."
  (declare (indent defun))
  (let ((modes (if (listp modes) modes (list modes))))
    (let ((setup-func
           (intern (concat "d/setup-"
                           (symbol-name (car modes))
                           "-mode"))))
      `(progn
         (defun ,setup-func ()
           ,@body)
         ,@(cl-loop for mode in modes collect
                    (let ((mode-hook (intern (concat (symbol-name mode)
                                                     "-mode-hook"))))
                      `(add-hook ',mode-hook #',setup-func)))))))

(defmacro d/with-eval-after-load (feature &rest body)
  (declare (indent defun))
  `(with-eval-after-load ,feature
     (condition-case-unless-debug err
         (progn
           ,@body)
       (error
        (display-warning 'config
                         (format "%s %s: %s " (symbol-name ,feature) "eval-after-load"
                                 (error-message-string err))
                         :error)))))
#+end_src
*** Other
#+begin_src emacs-lisp
(defun d/toggle-rlines ()
  "Toggle relative line numbers."
  (interactive)
  (if (eq display-line-numbers 'relative)
      (setq display-line-numbers t)
    (setq display-line-numbers 'relative)))


(define-minor-mode d/rlines-mode
  "Display relative line numbers for current buffer."
  nil
  " rl"
  nil
  (setq display-line-numbers nil)
  (when (and d/rlines-mode
             (not (minibufferp)))
    (setq display-line-numbers 'relative)
    (setq display-line-numbers-width
          (length
           (number-to-string
            (count-lines (point-min) (point-max)))))))

(define-globalized-minor-mode d/global-rlines-mode d/rlines-mode
  d/rlines-mode
  (lambda () (d/rlines-mode 1)))

(defmacro d/fbound-and-true-p (name &optional args)
  `(and (fboundp #',name)
        (apply #',name ,args)))

(defmacro time (name &rest body)
  (declare (indent defun))
  `(let ((s-time (current-time)))
     (prog1
         (progn
           ,@body)
       (message "`%s' execution took %.5f seconds."
                ,name
                (float-time (time-subtract (current-time) s-time))))))
#+end_src
** Packages
These are packages that I consider /absolutely essential/ to my emacs workflow,
or that enhance emacs at a deeper level than any regular mode. Or, I just had
to load them before other things.
*** [[https://github.com/tarsius/no-littering/][no-littering]]
#+begin_quote
  Help keeping ~/.emacs.d clean
#+end_quote
Usually, a bunch of crap is kept in your ~.emacs.d~ folder by both built-in emacs
features and external packages. This package sets up a convention to store
everything in either ~.emacs.d/var~ or ~.emacs.d/etc~.
#+begin_src emacs-lisp
(use-package no-littering
  :demand t
  :config
  (savehist-mode 1)
  (add-to-list 'savehist-additional-variables 'kill-ring)
  (save-place-mode 1)
  (setq auto-save-file-name-transforms
        `((".*" ,(no-littering-expand-var-file-name "auto-save/") t))))
#+end_src
*** [[https://github.com/purcell/exec-path-from-shell][exec-path-from-shell]]
#+begin_quote
  Make Emacs use the $PATH set up by the user's shell
#+end_quote
If you've ever had issues where emacs doesn't find your executables, this
package should fix them nicely.
#+begin_src emacs-lisp
(use-package exec-path-from-shell
  :defer 5
  :config
  (setq exec-path-from-shell-check-startup-files nil)
  (exec-path-from-shell-initialize))
#+end_src
*** [[https://github.com/noctuid/general.el][general]]
#+begin_quote
  More convenient key definitions in emacs
#+end_quote
That undersells it. The /most/ convenient key definitions in emacs.
#+begin_src emacs-lisp
(use-package general
  :after evil
  :recipe (:host github
           :repo "noctuid/general.el"
           :branch "buttercup")
  :config
  (general-evil-setup t t)

  (general-create-definer
   d/mode-leader-keys
   :states '(emacs normal visual motion insert)
   :non-normal-prefix "C-,"
   :prefix ",")

  (general-create-definer
   d/leader-keys
   :states '(emacs normal visual motion insert)
   :non-normal-prefix "C-SPC"
   :prefix "SPC"))
#+end_src
*** [[https://github.com/emacs-evil/evil][evil]]
#+begin_quote
  The extensible vi layer for Emacs.
#+end_quote
I really like Vim bindings. I originally learned Emacs bindings but there was
something really appealing about the simplicity and power of Vim bindings/modal
editing. So I went for it. Now I'll never go back.
**** Package
#+begin_src emacs-lisp
(use-package evil
  :demand t
  :general
  (nmap
   "S-SPC" 'org-agenda-list
   "-" 'negative-argument
   "\\" 'evil-window-next
   "C-_" 'keyboard-quit
   "C-/"  'keyboard-quit
   [escape]  'keyboard-quit)
  (:states '(insert replace visual)
   "C-_" 'evil-normal-state
   "C-/" 'evil-normal-state)
  (vmap [escape] 'keyboard-quit)
  :init
  (setq evil-want-C-u-scroll t
        evil-want-fine-undo t
        evil-search-module 'evil-search
        evil-lookup-func (lambda () (man (thing-at-point 'word))))
  :config
  (setq evil-insert-state-cursor '(bar . 1)
        evil-emacs-state-cursor '(bar . 1)
        evil-normal-state-tag  " N "
        evil-insert-state-tag  " I "
        evil-motion-state-tag  " M "
        evil-visual-state-tag  " V "
        evil-emacs-state-tag   " E "
        evil-replace-state-tag " R "
        evil-operator-state-tag " O ")

  (evil-ex-define-cmd "dtw" #'delete-trailing-whitespace)
  (evil-ex-define-cmd "buffers" #'ibuffer)

  (evil-mode 1))
#+end_src
**** Modes
#+begin_src emacs-lisp
(d/with-eval-after-load 'evil
  ;; Use Emacs state in these additional modes.
  (dolist (mode '(ag-mode
                  flycheck-error-list-mode
                  git-rebase-mode
                  eshell-mode
                  calc-mode
                  calc-trail-mode
                  sunshine-mode
                  term-mode
                  notmuch-tree-mode
                  profiler-report-mode))
    (add-to-list 'evil-emacs-state-modes mode))

  (setq evil-emacs-state-modes (delq 'ibuffer-mode evil-emacs-state-modes))
  (setq evil-emacs-state-modes (delq 'Custom-mode evil-emacs-state-modes))
  (setq evil-insert-state-modes (delq 'term-mode evil-insert-state-modes))

  ;; Use insert state in these additional modes.
  (dolist (mode '(magit-log-edit-mode org-capture-mode))
    (add-to-list 'evil-insert-state-modes mode))

  (add-to-list 'evil-buffer-regexps '("\\*Flycheck")))
#+end_src
*** [[https://github.com/abo-abo/swiper][ivy]]
#+begin_quote
  Ivy - a generic completion frontend for Emacs, Swiper - isearch with an
  overview, and more. Oh, man!
#+end_quote
A really nice search/completion system for emacs.
**** ivy
#+begin_src emacs-lisp
(use-package ivy
  :general
  (:keymaps 'ivy-minibuffer-map
   [escape] 'keyboard-escape-quit
   "C-/" 'keyboard-escape-quit
   [S-return] 'ivy-dispatching-done
   [C-return] 'ivy-immediate-done
   "C-j" 'ivy-next-line
   "C-k" 'ivy-previous-line
   [S-up] 'ivy-previous-history-element
   [S-down] 'ivy-next-history-element)
  (d/leader-keys
   "-" 'ivy-resume
   "bb" 'ivy-switch-buffer
   "bB" 'ivy-switch-buffer-other-window)
  :init
  (setq ivy-do-completion-in-region nil)
  :config
  (ivy-mode 1)

  (setq ivy-re-builders-alist '((swiper . ivy--regex-plus)
                                (t . ivy--regex-ignore-order)))
  (setq ivy-format-function 'ivy-format-function-arrow ; DAT NICE ARROW THOUGH aosenuth
        projectile-completion-system 'ivy
        ivy-use-virtual-buffers t ; Show recent files
        ivy-count-format ""
        ivy-extra-directories nil; '("../") ; ignore current folder and parent dir
        ivy-initial-inputs-alist '((man . "^"))
        ivy-switch-buffer-faces-alist '((dired-mode . ivy-subdir))
        recentf-max-saved-items 50
        ivy-use-selectable-prompt t))

(use-package ivy-hydra)
#+end_src
**** swiper
#+begin_src emacs-lisp
(use-package swiper
  :general
  (d/leader-keys
   "sm" 'swiper-multi
   "sS" 'swiper-all)
  :config (setq swiper-goto-start-of-match t))
#+end_src
**** counsel
#+begin_src emacs-lisp
(use-package counsel
  :general
  ("M-x" 'counsel-M-x
   "C-x C-f" 'counsel-find-file)
  (d/leader-keys
   "SPC"  'counsel-M-x
   "ff"  'counsel-find-file
   "fj"  'counsel-file-jump
   "fl"  'counsel-locate
   "hdF" 'counsel-describe-face
   "hdb" 'counsel-descbinds
   "hdf" 'counsel-describe-function
   "hdv" 'counsel-describe-variable
   "iu"  'counsel-unicode-char
   "sr"  'counsel-rg
   "ss"  'counsel-grep-or-swiper
   "y"   'counsel-yank-pop)
  :commands counsel-describe-face
  :config
  (when (eq system-type 'darwin)
    (setq counsel-locate-cmd 'counsel-locate-cmd-mdfind))

  (setq conusel-org-goto-display-style 'path
        counsel-org-goto-separator ": "
        counsel-org-goto-face-style 'org
        counsel-org-goto-display-todo t
        counsel-grep-base-command "grep -nEi '%s' %s"
        counsel-find-file-ignore-regexp (concat "\\(\\`\\.\\)"
                                                "\\|\\(\\.elc$\\)"
                                                "\\|\\(.pyc$\\)")
        counsel-yank-pop-separator "
  ─────────────────────────
")
  (counsel-mode 1)
  (defalias 'ag #'counsel-ag)
  (defalias 'locate #'counsel-locate)

  (define-advice counsel-yank-pop-action (:override (s) paste-after)
    "Paste text after point, which is consistent with evil-paste-after.

Source: https://git.io/vQKmf"
    (save-excursion
      (undo-boundary)
      (unless (eq (point) (point-max))
        (forward-char))
      (with-ivy-window
        (delete-region ivy-completion-beg
                       ivy-completion-end)
        (insert (substring-no-properties s))
        (setq ivy-completion-end (point))))
    (forward-char (length s)))

  (define-advice counsel--yank-pop-format-function (:override (cand-pairs) arrow-format)
    "Use the arrow format for counsel-yank-pop for consistency with
ivy-format-function-arrow.

Source: https://git.io/vQK0v"
    (ivy--format-function-generic
     (lambda (str)
       (let ((temp-list
              (split-string (counsel--yank-pop-truncate str) "\n" t)))
         (mapconcat 'identity
                    (append (list (concat "> " (car temp-list)))
                            (mapcar (lambda (s) (concat "  " s))
                                    (cdr temp-list)))
                    "\n")))
     (lambda (str)
       (mapconcat (lambda (s) (concat "  " s))
                  (split-string (counsel--yank-pop-truncate str) "\n" t)
                  "\n"))
     cand-pairs
     counsel-yank-pop-separator)))
#+end_src
*** [[https://github.com/Ilazki/prettify-utils.el][prettify-utils]]
#+begin_src emacs-lisp
(use-package prettify-utils
  :recipe (:host github
           :repo "Ilazki/prettify-utils.el"))
#+end_src
*** [[http://orgmode.org/][org-mode]]
#+begin_quote
  Org mode is for keeping notes, maintaining TODO lists, planning projects, and
  authoring documents with a fast and effective plain-text system.
#+end_quote
But really, it's life.

:NOTES:
*Note:* Getting emacs to run the latest version of org can be weird. Least I
didn't find a satisfactory solution for a while. Check out this [[http://sachachua.com/blog/2014/05/update-org-7-comes-emacs-org-8-configuration-better-exports/][blog post]] for
some advice on that (still relevant now-2016). In particular, make sure you
have something like:

#+begin_src emacs-lisp :tangle no
(package-initialize)
(setq package-enable-at-startup nil)
#+end_src

at the beginning of your init.el/emacs.d, or as in my case something like:

#+begin_src emacs-lisp :tangle no
;; (package-initialize)
(require 'package)
(setq package-enable-at-startup nil)
;; ---------------------------
;; my load-path settings here
;; ---------------------------
(package-initialize)
#+end_src

I actually don't think I tried the former option, but the latter simply worked
so I went with it.
:END:

**** Defaults
***** Files
#+begin_src emacs-lisp
(d/with-eval-after-load 'org
  (setq org-agenda-text-search-extra-files '(agenda-archives)
        org-agenda-files '("~/Dropbox/org/todo.org" "~/Dropbox/org/gcal.org")
        org-default-notes-file "~/Dropbox/org/todo.org"
        d/notes-file "~Dropbox/org/notes.org"
        org-directory "~/Dropbox/org"
        org-archive-location "~/Dropbox/org/archive.org::"
        org-mobile-inbox-for-pull "~/Dropbox/org/mobile.org"
        org-export-async-init-file
        (locate-user-emacs-file "lisp/org-async-init.el")))
#+end_src
***** Todo/agenda
#+begin_src emacs-lisp
(d/with-eval-after-load 'org
  (setq org-enforce-todo-dependencies t
        org-log-done (quote time)
        org-log-redeadline (quote time)
        org-log-reschedule (quote time)
        org-agenda-skip-scheduled-if-done t
        org-agenda-skip-deadline-if-done t
        org-agenda-hide-tags-regexp ".*"
        org-agenda-span 'week)

  (setq org-agenda-deadline-faces
        '((1.0 . org-warning)
          (0.5 . org-upcoming-deadline)
          (0.0 . '(:foreground "#A89984"))))

  (setq org-todo-keywords

config/emacs.d/config.el

(require 'package)
(setq package-enable-at-startup nil)
(add-to-list 'load-path
             (expand-file-name "lisp" user-emacs-directory))

;; Add directories in "lisp" folder
(let ((files (directory-files-and-attributes "~/.emacs.d/lisp" t)))
  (dolist (file files)
    (let ((filename (car file))
          (dir (nth 1 file)))
      (when (and dir
                 (not (string-suffix-p "." filename)))
        (add-to-list 'load-path (car file))))))

(add-to-list 'exec-path "/usr/local/bin")
(require 'init-utils)
(require 'init-elpa)

(package-initialize)

(unless (package-installed-p 'use-package)
  (package-refresh-contents)
  (package-install 'use-package))

(eval-when-compile
    (require 'use-package))

(setq user-full-name "Diego A. Mundo"
      user-mail-address "diegoamundo@gmail.com"
      calendar-location-name "Cambridge, MA")

(setq inhibit-splash-screen t
      inhibit-startup-message t
      inhibit-startup-echo-area-message t)
(menu-bar-mode 0)
(tool-bar-mode -1)
(toggle-scroll-bar -1)
(show-paren-mode 1)
(electric-pair-mode)
(add-hook 'minibuffer-setup-hook (lambda ()
                                   (electric-pair-mode)))
(setq display-time-load-average nil)
(setq cursor-type 'bar)
(fringe-mode 0)
(add-to-list 'default-frame-alist '(font . "Menlo-10"))
(setq scroll-conservatively 10000)
(setq-default tab-width 4)
(setq tab-stop-list (number-sequence 4 200 4))
(setq completion-cycle-threshold t)

; Store backups in .emacsbackups, autosaves in temp dir
(defvar backup-dir "~/.emacsbackups/")
(setq backup-directory-alist (list (cons "." backup-dir)))
(setq auto-save-file-name-transforms
      `((".*" ,temporary-file-directory t)))

;;; Custom Key Bindings ;;;
;; (global-set-key (kbd "<f8> <f1>")
;;              '(lambda () (interactive) (ansi-term "/usr/local/bin/ipython")))
;; (global-set-key (kbd "<f8> <f2>")
;;              '(lambda () (interactive) (ansi-term "/Users/diego/.virtualenvs/py2/bin/ipython")))
;; (global-set-key (kbd "<f9>") 'linum-mode)
;; (global-set-key (kbd "M-RET") 'python-shell-send-buffer)

;; Global stuff
(global-hl-line-mode 1)
;; (global-linum-mode 1)
(setq linum-delay t)
(global-auto-revert-mode t)
(setq whitespace-style '(face trailing))
(setq column-number-mode t)

(require 'init-global-functions)

(require 'init-evil)

;; Org prerequisites
(use-package visual-fill-column
     :ensure t)
(require 'init-org)

(use-package helm
  :ensure t
  :init
  (require 'helm-config)
  :config
  (use-package helm-descbinds
    ;; To describe keys in a nicer way
    :ensure t)
  (use-package helm-projectile
    ;; To use with projectile
    :ensure t
    :config
    (projectile-global-mode))
  ;; (use-package helm-ag
  ;;    :ensure t)
  (helm-mode 1)
  (helm-autoresize-mode t)
  (global-set-key (kbd "M-x") 'helm-M-x)  
  (global-set-key (kbd "<f1>") 'helm-find-files)  
  (global-set-key (kbd "<f2>") 'helm-mini)
  (setq helm-completion-in-region-fuzzy-match t)
  (setq helm-mode-fuzzy-match t)
  (setq helm-buffer-max-length 40)
  (setq helm-locate-command "mdfind -name %s %s") ; Use spotlight for search
  (global-set-key (kbd "M-y") 'helm-show-kill-ring))

(use-package magit
  :ensure t)

(use-package yasnippet
  ;; SNIPPETS!!!
  :ensure t
  :config
  (yas-global-mode 1))

(use-package flycheck
  ;; Pep8 check, basically
  :ensure t
  :config
  (global-set-key (kbd "<f11>") 'flycheck-mode))

(use-package company
  :ensure t
  :config
  (add-hook 'after-init-hook 'global-company-mode)
  (use-package company-jedi
    ;; Not sure this is actually working for me
    :ensure t
    :config
    (defun my/python-mode-hook ()
      (add-to-list 'company-backends 'company-jedi))
    (add-hook 'python-mode-hook 'my/python-mode-hook)))

(use-package ag
  ;; Silver searcher
  :ensure t
  :defer t
  :init
  (use-package wgrep-ag  
    ;; Guess I need this first
    :ensure t
    :commands (wgrep-ag-setup))
  :config
  (add-hook 'ag-mode-hook
            (lambda ()
              (wgrep-ag-setup)
              (define-key ag-mode-map (kbd "n") 'evil-search-next)
              (define-key ag-mode-map (kbd "N") 'evil-search-previous)))
  (setq ag-executable "/usr/local/bin/ag")
  (setq ag-highlight-search t)
  (setq ag-reuse-buffers t)
  (setq ag-reuse-window t))

(use-package fill-column-indicator
  :ensure t
  :init
  (setq-default fci-rule-column 79)
  (global-set-key (kbd "<f10>") 'fci-mode)
  (add-hook 'python-mode-hook 'fci-mode))

(use-package rainbow-delimiters
  :ensure t
  :init
  (add-hook 'python-mode-hook 'rainbow-delimiters-mode)
  (add-hook 'emacs-lisp-mode-hook 'rainbow-delimiters-mode))

(use-package highlight-parentheses
  ;; Make parenthesis I'm currently in stand out
  :ensure t)

(use-package highlight-numbers
  :ensure t
  :init
  (add-hook 'python-mode-hook 'highlight-numbers-mode))

(use-package imenu-anywhere
  ;; Imenu on steroids
  :ensure t
  :config
  (global-set-key (kbd "<f5>") 'imenu-anywhere))

(use-package spaceline
  ;; Similar to vim's powerline, this one looks clean
  ;; and 'just works', to an extent
  :ensure t
  :config
  (require 'spaceline-config)
  (spaceline-spacemacs-theme)
  (spaceline-helm-mode)
  (spaceline-toggle-minor-modes-off)
  (spaceline-toggle-battery-on)
  (spaceline-toggle-buffer-size-off)
  (setq spaceline-highlight-face-func 'spaceline-highlight-face-evil-state)
  (set-face-background 'spaceline-evil-normal "#afd700")
  (set-face-foreground 'spaceline-evil-normal "#005f00")
  (set-face-background 'spaceline-evil-insert "#0087af")
  (set-face-foreground 'spaceline-evil-insert "white")
  (set-face-background 'spaceline-evil-visual "#ff8700")
  (set-face-foreground 'spaceline-evil-visual "#870000"))

(use-package fancy-battery
  ;; Something something battery
  :ensure t
  :config
  (fancy-battery-mode)
  (setq fancy-battery-show-percentage t)
  (fancy-battery-update))

(use-package dictionary
  :ensure t)

(use-package sphinx-doc
  :ensure t
  :config
  (add-hook 'python-mode-hook
            (lambda ()
              (require 'sphinx-doc)
              (sphinx-doc-mode t))))

(use-package markdown-mode
  :ensure t)

(use-package markdown-preview-mode
  :ensure t)

(use-package csv-mode
  ;; I'll give this a shot
  :ensure t)

(use-package multi-term
  ;; Supposed to be nicer than ansi-term
  :ensure t)

;; (use-package smart-mode-powerline-theme
;;   :ensure t)

;; (use-package smart-mode-line
;;   :ensure t
;;   :config
;;   (setq sml/no-confirm-load-theme t)
;;   (setq sml/theme 'dark)
;;   (setq rm-whitelist '(""))
;;   (setq system-uses-terminfo nil)
;;   (sml/setup)
;;   (display-time-mode)
;;   (display-time-update)
;;   (fancy-battery-mode)
;;   (setq fancy-battery-show-percentage t))

;; (use-package elpy
;;   ;; Eh, I don't know...
;;   :ensure t
;;   :config
;;   (elpy-enable)
;;   (setq elpy-rpc-backend "jedi"))

;; (use-package yafolding
;;   ;; Man, good code folding is hard to come by in emacs
;;   ;; This one's ok, but there are a couple know issues that
;;   ;; don't quite make it worth it, I think.
;;   :ensure t
;;   :config
;;   (defun air--yafolding-kbd ()
;;  (local-set-key (kbd "C-c <up>") 'yafolding-hide-all)
;;  (local-set-key (kbd "C-c <down>") 'yafolding-show-all)
;;  (local-set-key (kbd "C-c <left>") 'yafolding-hide-element)
;;  (local-set-key (kbd "C-c <right>") 'yafolding-show-element)
;;  (local-set-key [C-tab] 'yafolding-toggle-element))
;;   (add-hook 'python-mode-hook 'yafolding-mode)
;;   (add-hook 'python-mode-hook 'air--yafolding-kbd))

;; (use-package jedi
;;   ;; Hasn't been working smoothly recently
;;   :ensure t
;;   :init
;;   (add-hook 'python-mode-hook 'jedi:setup)
;;   (setq jedi:complete-on-dot t))

;; (use-package auto-complete
;;   ;; Supposedly not as good as company mode
;;   :ensure t
;;   :config
;;   (global-auto-complete-mode t))

;; (use-package autopair
;;   :ensure t
;;   :config
;;   (autopair-global-mode))

(add-to-list 'auto-mode-alist '("\\.m$" . octave-mode))

;; Python stuff
(defun ipython ()
  (interactive)
  (ansi-term "/usr/local/bin/ipython"))
(defun ipython2()
  (interactive)
  (ansi-term "/Users/diego/.virtualenvs/py2/bin/ipython"))
(add-hook 'python-mode-hook 'highlight-parentheses-mode)
(add-hook 'python-mode-hook 'hs-minor-mode)
;;(add-hook 'python-mode-hook 'yafolding-mode)
(add-hook 'python-mode-hook 'yas-minor-mode)
(add-hook 'python-mode-hook
          (lambda () (set (make-local-variable 'comment-inline-offset) 2)))
(add-hook 'python-mode-hook (lambda () (setq tab-width 4)))
(add-hook 'python-mode-hook (lambda () (linum-mode 1)))
(setenv "PYTHONPATH" "/usr/local/bin/python3")

;; Emacs-lisp stuff
(defun my-lisp-mode-config ()
  (setq ac-sources '(ac-source-symbols ac-source-words-in-same-mode-buffers))
  (local-set-key (kbd "C-c <up>") 'hs-hide-all)
  (local-set-key (kbd "C-c <down>") 'hs-show-all)
  (local-set-key (kbd "C-c <left>") 'hs-hide-block)
  (local-set-key (kbd "C-c <right>") 'hs-show-block))

(add-hook 'emacs-lisp-mode-hook 'my-lisp-mode-config)
(add-hook 'emacs-lisp-mode-hook 'highlight-parentheses-mode)
(add-hook 'emacs-lisp-mode-hook 'hs-minor-mode)
(add-hook 'emacs-lisp-mode-hook (lambda () (linum-mode 1)))

;; Disable fci mode when autocomplete popup menu happens
(defun sanityinc/fci-enabled-p ()
    (and (boundp 'fci-mode) fci-mode))
(defvar sanityinc/fci-mode-suppressed nil)
(defadvice popup-create (before suppress-fci-mode activate)
  "Suspend fci-mode while popups are visible"
  (let ((fci-enabled (sanityinc/fci-enabled-p)))
    (when fci-enabled
      (set (make-local-variable 'sanityinc/fci-mode-suppressed) fci-enabled)
      (turn-off-fci-mode))))

(defadvice popup-delete (after restore-fci-mode activate)
  "Restore fci-mode when all popups have closed"
  (when (and sanityinc/fci-mode-suppressed
             (null popup-instances))
    (setq sanityinc/fci-mode-suppressed nil)
    (turn-on-fci-mode)))

(defface my-linum-hl
  `((t :inherit linum :background ,(face-background 'hl-line nil t)))
  "Face for the current line number."
  :group 'linum)

(defvar my-linum-format-string "%3d")

(add-hook 'linum-before-numbering-hook 'my-linum-get-format-string)

(defun my-linum-get-format-string ()
  (let* ((width (1+ (length (number-to-string
                             (count-lines (point-min) (point-max))))))
         (format (concat "%" (number-to-string width) "d ")))
    (setq my-linum-format-string format)))

(defvar my-linum-current-line-number 0)

(setq linum-format 'my-linum-format)

(defun my-linum-format (line-number)
  (propertize (format my-linum-format-string line-number) 'face
              (if (eq line-number my-linum-current-line-number)
                  'my-linum-hl
                'linum)))

(defadvice linum-update (around my-linum-update)
  (let ((my-linum-current-line-number (line-number-at-pos)))
    ad-do-it))
(ad-activate 'linum-update)
(add-hook 'term-mode-hook (lambda ()
                            (setq-local global-hl-line-mode
                                        nil)))
